fe/fe-core/src/main/java/com/starrocks/analysis/CreateFunctionStmt.java

```diff
@@ -215,93 +349,115 @@ private void computeObjectChecksum() throws IOException, NoSuchAlgorithmExceptio
         }
-        function = builder.build();
+    }
+
+    private void analyzeStarrocksJarUdf() throws AnalysisException {
+        checkStarrocksJarUdfClass();
+        function = ScalarFunction.createUdf(
+                functionName, argsDef.getArgTypes(),
+                returnType.getType(), argsDef.isVariadic(), TFunctionBinaryType.SRJAR,
+                objectFile, udfClass.getCanonicalName(), "", "");
         function.setChecksum(checksum);
     }
 
-    private void checkStarrocksJarUdfType(Type type, Class ptype, String pname) throws AnalysisException {
-        if (!(type instanceof ScalarType)) {
-            throw new AnalysisException("UDF does not support non-scalar type: " + type);
+    private void checkStarrocksJarUdafStateClass() throws AnalysisException {
+        // Check internal State class
+        // should be public & static.
+        Class stateClass = udfStateClass.clazz;
+        if (!Modifier.isPublic(stateClass.getModifiers()) ||
+                !Modifier.isStatic(stateClass.getModifiers())) {
+            throw new AnalysisException(
+                    String.format("UDAF '%s' should have one public & static 'State' class",
+                            udfClass.getCanonicalName()));
         }
-        ScalarType scalarType = (ScalarType) type;
-        Class cls = PrimitiveTypeToJavaClassType.get(scalarType.getPrimitiveType());
-        if (cls == null) {
-            throw new AnalysisException("UDF does not support type: " + scalarType);
-        }
-        if (!cls.equals(ptype)) {
-            throw new AnalysisException(String.format("UDF %s[%s] type does not match %s", pname,
-                    ptype.getCanonicalName(), cls.getCanonicalName()));
+        {
+            // long serializeLength();
+            Method method = udfStateClass.getMethod(SERIALIZE_LENGTH_METHOD_NAME, true);
+            udfStateClass.checkMethodNonStaticAndPublic(method);
+            udfStateClass.checkReturnJavaType(method, long.class);
+            udfStateClass.checkArgumentCount(method, 0);
         }
     }
 
-    private void checkStarrocksJarUdfMethod(Method method) throws AnalysisException {
-        String name = method.getName();
-        boolean checked = true;
-        if (EVAL_METHOD_NAME.equals(name)) {
-            Class retType = method.getReturnType();
-            checkStarrocksJarUdfType(returnType.getType(), retType, "Return");
-            if (method.getParameters().length != argsDef.getArgTypes().length) {
-                throw new AnalysisException(String.format("UDF '%s' parameter count does not match", name));
-            }
-            for (int i = 0; i < method.getParameters().length; i++) {
-                Parameter p = method.getParameters()[i];
-                checkStarrocksJarUdfType(argsDef.getArgTypes()[i], p.getType(), p.getName());
-            }
-        } else {
-            checked = false;
+    private void checkStarrocksJarUdafClass() throws AnalysisException {
+        {
```

Comment:
Don't need define multi code block, reuse method is OK, like:
```
Method method;
method = udfClass.getMethod(CREATE_METHOD_NAME, true);
......
......

method = udfClass.getMethod(DESTROY_METHOD_NAME, true);
```

Reply:
Yes. I want to split to multi-blocks to emphasize checking of each function.